core/src/input_manager.cpp

```diff
@@ -0,0 +1,205 @@
+    return m_padBindings[(int)pad][(int)button] && !m_holdPadBindings[(int)pad][(int)button];
+}
+
+bool input_manager::m_padButtonReleased(PadCode pad, PadButtonCode button) {
+    return m_padBindings[(int)pad][(int)button] && !m_holdPadBindings[(int)pad][(int)button];
+}
+
+float input_manager::m_getPadAxis(PadCode pad, JoyStickCode joy_stick) {
+    return m_joyStickBindings[(int)pad][(int)joy_stick];
+}
+
+void input_manager::m_updatePads(GLFWwindow* glfwWindow) {
+    for (int i = 0; i < (int)PadCode::Last; i++) {
+        if (!m_padPresent((PadCode)i))
+            continue;
+
+        int count;
+        const float* axis = glfwGetJoystickAxes(i, &count);
+        if (!axis)
+            continue;
+
+        if (count > (int)JoyStickCode::Last)
+            count = (int)JoyStickCode::Last;
+        for (int n = 0; n < count; ++n)
+            m_joyStickBindings[i][n] = axis[n];
+
+        GLFWgamepadstate gameState;
+        if (glfwGetGamepadState(i, &gameState)) {
+            for (int j = 0; j < (int)PadButtonCode::Last; j++) {
+                bool padPressed = gameState.buttons[j];
+                m_padBindings[i][j] = padPressed;
+            }
+        } else if (glfwJoystickPresent(i)) {
+            const unsigned char* buttonPressed = glfwGetJoystickButtons(i, &count);
+            if (!buttonPressed)
+                continue;
+            for (int j = 0; j < count; ++j) {
+                if (count > 0)
+                    m_padBindings[i][(int)PadButtonCode::A] = buttonPressed[0];
+                if (count > 1)
+                    m_padBindings[i][(int)PadButtonCode::B] = buttonPressed[1];
+                if (count > 3)
+                    m_padBindings[i][(int)PadButtonCode::X] = buttonPressed[3];
+                if (count > 4)
+                    m_padBindings[i][(int)PadButtonCode::Y] = buttonPressed[4];
+                if (count > 6)
+                    m_padBindings[i][(int)PadButtonCode::LeftBumper] = buttonPressed[6];
+                if (count > 7)
+                    m_padBindings[i][(int)PadButtonCode::RightBumper] = buttonPressed[7];
+                if (count > 11)
+                    m_padBindings[i][(int)PadButtonCode::Start] = buttonPressed[11];
+                if (count > 13)
+                    m_padBindings[i][(int)PadButtonCode::LeftThumb] = buttonPressed[13];
+                if (count > 14)
+                    m_padBindings[i][(int)PadButtonCode::RightThumb] = buttonPressed[14];
+                if (count > 15)
+                    m_padBindings[i][(int)PadButtonCode::DPadUp] = buttonPressed[15];
+                if (count > 16)
+                    m_padBindings[i][(int)PadButtonCode::DPadRight] = buttonPressed[16];
+                if (count > 17)
+                    m_padBindings[i][(int)PadButtonCode::DPadDown] = buttonPressed[17];
+                if (count > 18)
+                    m_padBindings[i][(int)PadButtonCode::DPadLeft] = buttonPressed[18];
+            }
```

Comment:
This is code from 5 years ago, I know it is mine, but it just hurts. Please make it pretty.